src/raw/mod.rs

```diff
@@ -4161,45 +3642,17 @@ impl<T> Iterator for RawIterHash<T> {
     type Item = Bucket<T>;
 
     fn next(&mut self) -> Option<Bucket<T>> {
-        unsafe {
-            match self.inner.next() {
-                Some(index) => {
-                    // Can't use `RawTable::bucket` here as we don't have
-                    // an actual `RawTable` reference to use.
-                    debug_assert!(index <= self.inner.bucket_mask);
-                    let bucket = Bucket::from_base_index(self.inner.ctrl.cast(), index);
-                    Some(bucket)
-                }
-                None => None,
-            }
-        }
-    }
-}
-
-impl Iterator for RawIterHashInner {
-    type Item = usize;
-
-    fn next(&mut self) -> Option<Self::Item> {
-        unsafe {
-            loop {
-                if let Some(bit) = self.bitmask.next() {
-                    let index = (self.probe_seq.pos + bit) & self.bucket_mask;
-                    return Some(index);
-                }
-                if likely(self.group.match_empty().any_bit_set()) {
-                    return None;
-                }
-                self.probe_seq.move_next(self.bucket_mask);
-
-                // Can't use `RawTableInner::ctrl` here as we don't have
-                // an actual `RawTableInner` reference to use.
-                let index = self.probe_seq.pos;
-                debug_assert!(index < self.bucket_mask + 1 + Group::WIDTH);
-                let group_ctrl = self.ctrl.as_ptr().add(index).cast();
-
-                self.group = Group::load(group_ctrl);
-                self.bitmask = self.group.match_tag(self.tag_hash).into_iter();
+        match self.inner.next() {
+            Some((_, ptr)) => {
+                // FIXME: this is the worst API ever oh no oh dear I hate it
+                // FIXME: use NonNull::add if MSRV > 1.80.0. hopefully this is
+                //   deleted before then…
+                // SAFETY: Bucket wants our pointer to be offset by one, so,
+                //   make it so…
+                let ptr = unsafe { NonNull::new_unchecked(ptr.as_ptr().add(1)) };
+                Some(Bucket { ptr })
```

Comment:
Shout outs to miri for helping me find this issue.

So, okay. I didn't want to fix *all* usages of `Bucket` in this PR and make them all actually put the pointer in the *correct* place since that's a big change. But I'm also not changing `ProbeItems` to return a pointer to somewhere *other* than where the item is.

If it's actually a big enough deal for performance, I'll try and fix the bucket API to be less bad. But that API is on the chopping block and hopefully I can get to it before we want to do a release anyway.